From 455d3091c2aa1869ec58a89744b0c686e4a5c9d6 Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Fri, 25 Feb 2022 11:57:46 +0530
Subject: [PATCH] ath12k: Add support for VHT in 2G

In 2G,VHT supported MCS Rx and Tx Map is set to unsupported in
VHT capabilities IE field of association response which makes the
STA to deny association in VHT/HE mode.
Add support to setup Tx and Rx VHT MCS map to VHT capability
which will be advertised in IE.

Enable VHT supported capability in 2G during STA assocation.

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>

Index: backports-20220209-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/mac.c
===================================================================
--- backports-20220209-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/mac.c
+++ backports-20220209-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/mac.c
@@ -4355,6 +4355,8 @@ static void ath12k_mac_setup_ht_vht_cap(
 			*ht_cap_info = ht_cap;
 		band->ht_cap = ath12k_create_ht_cap(ar, ht_cap,
 						    rate_cap_rx_chainmask);
+		band->vht_cap = ath12k_create_vht_cap(ar, rate_cap_tx_chainmask,
+						    rate_cap_rx_chainmask);
 	}
 
 	if (cap->supported_bands & WMI_HOST_WLAN_5G_CAP && !ar->supports_6ghz) {
Index: backports-20220209-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/wmi.c
===================================================================
--- backports-20220209-5.4.164-f40abb4788.orig/drivers/net/wireless/ath/ath12k/wmi.c
+++ backports-20220209-5.4.164-f40abb4788/drivers/net/wireless/ath/ath12k/wmi.c
@@ -412,6 +412,8 @@ ath12k_pull_mac_phy_cap_svc_ready_ext(st
 	 * handled.
 	 */
 	if (mac_phy_caps->supported_bands & WMI_HOST_WLAN_2G_CAP) {
+		pdev_cap->vht_cap = mac_phy_caps->vht_cap_info_2g;
+		pdev_cap->vht_mcs = mac_phy_caps->vht_supp_mcs_2g;
 		pdev_cap->tx_chain_mask = mac_phy_caps->tx_chain_mask_2g;
 		pdev_cap->rx_chain_mask = mac_phy_caps->rx_chain_mask_2g;
 	} else if (mac_phy_caps->supported_bands & WMI_HOST_WLAN_5G_CAP) {
Index: backports-20220209-5.4.164-f40abb4788/net/mac80211/mlme.c
===================================================================
--- backports-20220209-5.4.164-f40abb4788.orig/net/mac80211/mlme.c
+++ backports-20220209-5.4.164-f40abb4788/net/mac80211/mlme.c
@@ -5028,8 +5028,6 @@ static int ieee80211_prep_channel(struct
 	struct ieee802_11_elems *elems;
 	const struct cfg80211_bss_ies *ies;
 	int ret;
-	u32 i;
-	bool have_80mhz;
 
 	rcu_read_lock();
 
@@ -5127,22 +5125,6 @@ static int ieee80211_prep_channel(struct
 			ifmgd->flags |= IEEE80211_STA_DISABLE_HE;
 	}
 
-	/* Allow VHT if at least one channel on the sband supports 80 MHz */
-	have_80mhz = false;
-	for (i = 0; i < sband->n_channels; i++) {
-		if (sband->channels[i].flags & (IEEE80211_CHAN_DISABLED |
-						IEEE80211_CHAN_NO_80MHZ))
-			continue;
-
-		have_80mhz = true;
-		break;
-	}
-
-	if (!have_80mhz) {
-		sdata_info(sdata, "80 MHz not supported, disabling VHT\n");
-		ifmgd->flags |= IEEE80211_STA_DISABLE_VHT;
-	}
-
 	if (sband->band == NL80211_BAND_S1GHZ) {
 		s1g_oper = elems->s1g_oper;
 		if (!s1g_oper)
Index: backports-20220209-5.4.164-f40abb4788/net/mac80211/vht.c
===================================================================
--- backports-20220209-5.4.164-f40abb4788.orig/net/mac80211/vht.c
+++ backports-20220209-5.4.164-f40abb4788/net/mac80211/vht.c
@@ -121,7 +121,6 @@ ieee80211_vht_cap_ie_to_sta_vht_cap(stru
 	struct ieee80211_sta_vht_cap *vht_cap = &sta->sta.vht_cap;
 	struct ieee80211_sta_vht_cap own_cap;
 	u32 cap_info, i;
-	bool have_80mhz;
 
 	memset(vht_cap, 0, sizeof(*vht_cap));
 
@@ -131,20 +130,6 @@ ieee80211_vht_cap_ie_to_sta_vht_cap(stru
 	if (!vht_cap_ie || !sband->vht_cap.vht_supported)
 		return;
 
-	/* Allow VHT if at least one channel on the sband supports 80 MHz */
-	have_80mhz = false;
-	for (i = 0; i < sband->n_channels; i++) {
-		if (sband->channels[i].flags & (IEEE80211_CHAN_DISABLED |
-						IEEE80211_CHAN_NO_80MHZ))
-			continue;
-
-		have_80mhz = true;
-		break;
-	}
-
-	if (!have_80mhz)
-		return;
-
 	/*
 	 * A VHT STA must support 40 MHz, but if we verify that here
 	 * then we break a few things - some APs (e.g. Netgear R6300v2
